### ISSUE TITLE
Is the training unstable？

### ISSUE THREAD

**Issue by 978749951** [freesunshine0316/neural-graph-to-seq-mp#9]
Hi@freesunshine0316 
     Thanks for sharing  the pre-trained model with only with gold and the decoding script. As you described,  It reports a test BLEU score of 23.3. However, I used the same  configuration to train model by myself and only get a  test BLEU score of 21.6. Have you ever  found the same phenomenon? I guess that the result is unstable because the sparse of data? If not, could you give me  any advice?

**Comment by freesunshine0316** [comment]
Did you decrease the learning rate after several numbers of epochs? Both Konstas et al and I found it helpful and used it to achieve the reported BLEU scores.  If your initial learning rate is 1e-3, I  suggest you train for another 5 or 6 epochs with a lower learning rate, such as 8e-4.

**Comment by Fije** [comment]
Hi, is there a way I can use a pre-trained model and finetune it with my own data?

**Comment by vivald99** [comment]
Hi @978749951 , Could you help me with an issue?

Did you get the test BLEU score of 21.6 using a model trained with the same configuration ([G2S.gold.config.json](https://www.cs.rochester.edu/~lsong10/downloads/model_gold.tgz))? I also trained the model using that file and decreasing the learning rate as cited in the README, but I got very low test BLEU score: 14.13. 
When I use the available pre-trained model, I get test BLEU score of 21.58 (like you).

I'm using python 2.7.15 with tensorflow 1.4.1, as cited in the README.

Did you just use the G2S.gold.config.json file, or did something else?

**Comment by freesunshine0316** [comment]
@978749951 @vivald99 I just realized I forgot to include the pretrained embeddings in the gold model, but they are in the silver_2m model. To achieve the reported number, we have to use pretrained embeddings (and DO NOT update them during training), as the gold data is too small. I just incorporate my pretrained embeddings into the gold model.

**Comment by vivald99** [comment]
Hi @freesunshine0316 , thank you very much for your response!

I used the pretrained word embeddings for the silver_2m model. However, I didn't get test BLEU scores similar as yours (your and @978749951 's). Only with your pre-trained model, I got similar result (21.58).

I'm trying to find where I'm doing something wrong.

This was the config file that I used:

{"dropout_rate": 0.1, "suffix": "gold", "test_path": "data/dev.json", "highway_layer_num": 1, "max_in_neigh_num": 2, "optimize_type": "adam", "max_answer_len": 100, "with_highway": true, "max_epochs": 50, "reward_type": "bleu", "max_node_num": 200, "word_vec_path": "data/vectors_RwNN.txt.st", "max_char_per_word": 20, "config_path": "config.json", "attention_vec_size": 300, "lambda_l2": 0.001, "fix_word_vec": true, "cov_loss_wt": 0.1, "num_syntax_match_layer": 9, "model_dir": "logs_g2s", "edgelabel_dim": 50, "way_init_decoder": "root", "with_char": true, "finetune_path": "", "learning_rate": 0.001, "batch_size": 20, "char_dim": 50, "compress_input_dim": 300, "max_out_neigh_num": 10, "isLower": true, "compress_input": true, "attention_type": "hidden_embed", "pointer_gen": true, "gen_hidden_size": 300, "CE_loss": false, "char_lstm_dim": 100, "best_accu": 0.46643160917451204, "use_coverage": true, "num_softmax_samples": 100, "beam_size": 5, "neighbor_vector_dim": 300, "mode": "ce_train", "train_path": "data/training.json", "min_answer_len": 0, "generate_config": false}

**Comment by freesunshine0316** [comment]
@vivald99 You mean you only use it for decoding? That's wired. What script do you use for evaluation? I use multi-bleu.perl from Moses. Also, remember to remove the "\<\/s\>" tags before evaluation.

**Comment by vivald99** [comment]
@freesunshine0316 , I used your pre-trained model only for decoding. But I also trained a model using your config file.

I wrote a script to get the result file generated by your model and transform it to be used in multi-bleu.perl:
```
import sys
from os import system
import numpy as np

f = open(sys.argv[1],'r')
refs = []
decs = []

while 1:
 line = f.readline()
 if not line:
  break
 line = f.readline()
 ref = line.strip()
 line = f.readline()
 dec = line.replace('</s>','').strip()
 line = f.readline()
 line = f.readline()
 refs.append(ref)
 decs.append(dec)

with open('hypothesis.txt', 'w') as output_file:
 for v in decs:
  output_file.write(v + "\n")

with open('reference.txt', 'w') as reference_file:
 for v in refs:
  reference_file.write(v + "\n")

system('./multi-bleu.perl -lc reference.txt < hypothesis.txt > bleu.out')

f = open('bleu.out','r')
print(f.readlines())
```

With your pre-trained model (gold) I got a BLEU score in the testset of 21.58 (using the script above). With a model that I trained I got 14.13. Maybe, I got these results because of a bad generation of json files for training, dev and test. I'm investigating that.

**Comment by freesunshine0316** [comment]
Why deleting commas? You can do something like this:

```
output_dict = {}
ref_dict = {}
for i,line in enumerate(open(sys.argv[1],'rU')):
    if i%5 == 0:
        id = line.strip()
    elif i%5 == 1:
        ref = line.strip().lower().split()
        ref_dict[id] = ref
    elif i%5 == 2:
        rst = line.strip().replace('</s>','',10).split()
        output_dict[id] = rst
```

**Comment by vivald99** [comment]
I was not deleting commas. Github deleted a part of the code. I edited. Thank you for your help.

**Comment by vivald99** [comment]
Hi @freesunshine0316 . I have the following code (I changed your code a little bit) to transform the files of the testset (LDC2015E86_DEFT_Phase_2_AMR_Annotation_R1) to a json file:
```
# first argument -> folder of amr files
# second argument -> generated file

import sys
import glob
import json

ids = []
id_dict = {}
amrs = []
sents = []

print sys.argv
for f in glob.glob(sys.argv[1]+"/*.txt"):
 amr_str = ''
 for line in open(f,'rU'):
  if line.startswith('#'):
   if line.startswith('# ::id'):
    id = line.lower().strip().split()[2]
    ids.append(id)
    if(id in id_dict):
     print len(ids),id
    id_dict[id] = len(ids)-1
   if line.startswith('# ::snt'):
    sent = line.split('# ::snt')[1].strip()
    sents.append(sent)
   continue
  line = line.strip()
  if("#" in line):
   print line
   line = line.replace("#","")
  if line == '':
   if amr_str != '':
    amrs.append(amr_str.strip())
    amr_str = ''
  else:
   amr_str = amr_str + line + ' '
 if amr_str != '':
  amrs.append(amr_str.strip())
  amr_str = ''


print len(amrs)
print len(ids)
print len(id_dict)


nf = open(sys.argv[2],'w')
nfc = open(sys.argv[2]+'.comp','w')
json_data = []

for id in ids:
 nf.write(amrs[id_dict[id]]+'\n')
 j = {}
 j['id'] = id
 j['id_rev'] = id_dict[id]
 j['sent'] = sents[id_dict[id]]
 j['amr_original'] = amrs[id_dict[id]]
 json_data.append(j)

print len(json_data)

json.dump(json_data,nfc)
                      
```

- After that I run the anonDeAnon_java.sh file to generate the anonymized amr data. 
- I run the following code to generate the json file (test.json), combining the anonymized amr with the json data:
```
import json
import sys

data_file = open(sys.argv[1],'r')
anom_file = open(sys.argv[2],'r')
an_amr = []
for line in anom_file:
 an_amr.append(line.strip())

data_comp = json.load(data_file)
print(len(data_comp))
print(len(an_amr))
new_data_comp = []

for id,amr in enumerate(an_amr):
 if amr == 'FAILED_TO_PARSE':
  print amr
  continue
 for d in data_comp:
  if(d['id_rev'] == id):
   d['amr'] = amr
   new_data_comp.append(d)

print (len(new_data_comp))

with open(sys.argv[3],'w') as j_file:
 json.dump(new_data_comp,j_file)
```
- So I use this json file as input to your pre-trained model
- Finally, I ran the multi bleu script (that I showed in the previous comment) to evaluate the generation and I got the following line:
['BLEU = 18.93, 50.6/24.4/13.4/7.7 (BP=1.000, ratio=1.014, hyp_len=27484, ref_len=27096)\n']

It should be something near a BLEU score of 23, right? Do you have any guess about why I can't get good results?

**Comment by freesunshine0316** [comment]
@vivald99 I can't see your outputs for more investigation, so I uploaded my eval_script, test outputs for gold and my test.json files. You can check them if there can be any help.

**Comment by vivald99** [comment]
@freesunshine0316 thank you very much. I will evaluate them and report to you.

**Comment by vivald99** [comment]
Hi @freesunshine0316 . Thank you for the files. I think I found the problem. It has to do with the tokenizer used in the sentences. 

1 example:
Your file:
` i 'm 21 yrs old ( turn 22 this year ) , do n't have much time left for failing need to do something .`
My file:
`i 'm 21yrs old ( turn 22 this year ) , do n't have much time left for failing need to do something .`

2 example:
Your file:
`its illegal to pay certain amounts in loads of coins . such as you can only pay something like 30 pence in 1 ps`
My file:
`its illegal to pay certain amounts in loads of coins . such as you can only pay something like 30pence in 1ps`

My tokenizer (nltk) generate tokens a little different from yours. So I think that to use my data, I need to generate a new vocab and train the model again. 

I was using your vocab with my generated data (for test) and your pre-trained model. I think I did not get your results because of the vocab of my generated test.json file.

Using your test.json with your pre-trained model I got the same bleu score reported by you. Which tokenizer did you use? Do you remember?

Thank you again for your help.

**Comment by freesunshine0316** [comment]
I used the tokenizer by stanford-core-nlp, which is based on Penn Tree Bank style.
https://www.google.com/search?client=firefox-b-1-d&q=ptb_tokenizer
